src/main/java/org/tikv/raw/RawKVClient.java

```diff
@@ -190,6 +189,123 @@ public void delete(ByteString key) {
     }
   }
 
+  private class Batch {
+    private long regionId;
+    private List<ByteString> keys;
+    private List<ByteString> values;
+
+    public Batch(long regionId, List<ByteString> keys, List<ByteString> values) {
+      this.regionId = regionId;
+      this.keys = keys;
+      this.values = values;
+    }
+  }
+
+  /**
+   * Append batch to list and split them according to batch limit
+   *
+   * @param batches a grouped batch
+   * @param regionId region ID
+   * @param keys keys
+   * @param values values
+   * @param limit batch max limit
+   */
+  private void appendBatches(
+      List<Batch> batches,
+      long regionId,
+      List<ByteString> keys,
+      List<ByteString> values,
+      int limit) {
+    List<ByteString> tmpKeys = new ArrayList<>();
+    List<ByteString> tmpValues = new ArrayList<>();
+    for (int i = 0; i < keys.size(); i++) {
+      if (i >= limit) {
+        batches.add(new Batch(regionId, tmpKeys, tmpValues));
+        tmpKeys.clear();
+        tmpValues.clear();
+      }
+      tmpKeys.add(keys.get(i));
+      tmpValues.add(values.get(i));
+    }
+    if (!tmpKeys.isEmpty()) {
+      batches.add(new Batch(regionId, tmpKeys, tmpValues));
+    }
+  }
+
+  /**
+   * Group by list of keys according to its region
+   *
+   * @param keys keys
+   * @return a mapping of keys and their regionId
+   */
+  private Map<Long, List<ByteString>> groupKeysByRegion(List<ByteString> keys) {
+    Map<Long, List<ByteString>> groups = new HashMap<>();
+    TiRegion lastRegion = null;
+    for (ByteString key : keys) {
+      if (lastRegion == null || !lastRegion.contains(key)) {
+        lastRegion = regionManager.getRegionByKey(key);
+      }
+      groups.computeIfAbsent(lastRegion.getId(), k -> new ArrayList<>()).add(key);
+    }
+    return groups;
+  }
+
+  static Map<ByteString, ByteString> mapKeysToValues(
+      List<ByteString> keys, List<ByteString> values) {
+    Map<ByteString, ByteString> map = new HashMap<>();
+    for (int i = 0; i < keys.size(); i++) {
+      map.put(keys.get(i), values.get(i));
+    }
+    return map;
+  }
+
+  /**
+   * Send batchPut request concurrently
+   *
+   * @param backOffer current backOffer
+   * @param batches list of batch to send
+   */
+  private void sendBatchPut(BackOffer backOffer, List<Batch> batches) {
+    for (Batch batch : batches) {
+      completionService.submit(
+          () -> {
+            Pair<TiRegion, Metapb.Store> pair =
+                regionManager.getRegionStorePairByRegionId(batch.regionId);
+            RegionStoreClient client = RegionStoreClient.create(pair.first, pair.second, session);
+            List<Kvrpcpb.KvPair> kvPairs = new ArrayList<>();
+            for (int i = 0; i < batch.keys.size(); i++) {
```

Comment:
Can we directly use List<KvPair> inside of batch? It's kind of meaningless to copy twice.

Reply:
done.